<a href="https://colab.research.google.com/github/RailyaR/project_5/blob/main/Tariff_recommendation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

**Описание данных**

**Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:**
    
*сalls* — количество звонков,

*minutes* — суммарная длительность звонков в минутах,

*messages* — количество sms-сообщений,

*mb_used* — израсходованный интернет-трафик в Мб,

*is_ultra* — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

## Откройте и изучите файл

In [ ]:
import pandas as pd
from sklearn.tree import DecisionTreeRegressor
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
df = pd.read_csv('/content/users_behavior.csv')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


<div class="alert alert-info">
<b>🎯 Комментарий : </b> 
<br> Данные выгружены:
<br>  Признак, который нужно предсказать, — целевой: в нашей задаче это is_ultra. И речь идёт о бинарной (двоичной) классификации.
    
</div>

In [ ]:
df.is_ultra.value_counts()

0    2229
1     985
Name: is_ultra, dtype: int64

In [ ]:
is_ultra_0 = 2229
is_ultra_1 = 985

## Разбейте данные на выборки

Разделите исходные данные на обучающую, валидационную и тестовую выборки.

In [ ]:
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']
features_train,features_valid,target_train,target_valid =  train_test_split(features,target, test_size=0.4, random_state=12345, stratify=target)
features_test, features_valid,target_test,target_valid =  train_test_split(features_valid,target_valid, test_size=0.5, random_state=12345,stratify=target_valid)

## Исследуйте модели

In [ ]:
best_depth = 0
best_accuracy = 0

for depth in range(1,40):
    model = DecisionTreeClassifier(random_state=1234567, max_depth=depth)
    model.fit(features_train,target_train)
    predictions_valid = model.predict(features_valid) 
    result = accuracy_score(target_valid,predictions_valid)
    print('Глубина дерева', depth,'Точность',result)
    if result > best_accuracy:
        best_depth = depth
        best_accuracy =result

print('Лучшая глубина дерева', best_depth,'Лучшая точность', best_accuracy)

Глубина дерева 1 Точность 0.7293934681181959
Глубина дерева 2 Точность 0.7682737169517885
Глубина дерева 3 Точность 0.7993779160186625
Глубина дерева 4 Точность 0.7744945567651633
Глубина дерева 5 Точность 0.8118195956454122
Глубина дерева 6 Точность 0.7978227060653188
Глубина дерева 7 Точность 0.7962674961119751
Глубина дерева 8 Точность 0.7822706065318819
Глубина дерева 9 Точность 0.7931570762052877
Глубина дерева 10 Точность 0.7947122861586314
Глубина дерева 11 Точность 0.7853810264385692
Глубина дерева 12 Точность 0.7729393468118196
Глубина дерева 13 Точность 0.76049766718507
Глубина дерева 14 Точность 0.7402799377916018
Глубина дерева 15 Точность 0.7433903576982893
Глубина дерева 16 Точность 0.7356143079315708
Глубина дерева 17 Точность 0.7231726283048211
Глубина дерева 18 Точность 0.7231726283048211
Глубина дерева 19 Точность 0.7200622083981337
Глубина дерева 20 Точность 0.7153965785381027
Глубина дерева 21 Точность 0.7200622083981337
Глубина дерева 22 Точность 0.71850699844479
Г

In [ ]:
best_model_TreeClass = DecisionTreeClassifier(random_state=1234567, max_depth=5)
best_model_TreeClass.fit(features_train,target_train)
predictions_valid = best_model_TreeClass.predict(features_valid) 
result = accuracy_score(target_valid,predictions_valid)
print('Глубина дерева', 5,'Точность',result)

Глубина дерева 5 Точность 0.8118195956454122


Лучшая глубина дерева 5 Лучшая точность 0.8118195956454122

In [ ]:
model = DecisionTreeClassifier(random_state=1234567, max_depth=5,criterion="entropy")
model.fit(features_train,target_train)
predictions_valid = model.predict(features_valid) 
result = accuracy_score(target_valid,predictions_valid)
print('Глубина дерева', 5,'Точность',result)

Глубина дерева 5 Точность 0.8087091757387247


<div class="alert alert-info">
<b>🎯 Комментарий : </b> 
<br> Данные выгружены:
<br>  Глубина дерева 5 Точность 0.8087091757387247, при гиперпараметре criterion="entropy". Точность уменьшилась. Поэтому данный гиперпараметр не подходит.
    
</div>

In [ ]:
model = DecisionTreeClassifier(random_state=1234567, max_depth=5,splitter='random')
model.fit(features_train,target_train)
predictions_valid = model.predict(features_valid) 
result = accuracy_score(target_valid,predictions_valid)
print('Глубина дерева', 5,'Точность',result)


Глубина дерева 5 Точность 0.7465007776049767


 
<div class="alert alert-info">
<b>🎯 Комментарий : </b> 
<br> Глубина дерева 5; Точность 0.7465007776049767. При гиперпараметре splitter='random',точность тоже уменьшилась, поэтому не подходит. 

    
</div>

In [ ]:
best_est = 0
best_accuracy = 0


for est in range(1, 50):
    model = RandomForestClassifier(random_state=12345, n_estimators=est)
    model.fit(features_train,target_train)
    predictions_valid = model.predict(features_valid) 
    result = accuracy_score(target_valid,predictions_valid)
    print('Количество деревьев', est,'Точность',result)
    if result > best_accuracy:
            best_est = est
            best_accuracy =result
            best_depth_rf = depth
print('Количество деревьев', best_est,'Лучшая точность', best_accuracy)

Количество деревьев 1 Точность 0.7402799377916018
Количество деревьев 2 Точность 0.7558320373250389
Количество деревьев 3 Точность 0.7480559875583204
Количество деревьев 4 Точность 0.7744945567651633
Количество деревьев 5 Точность 0.7713841368584758
Количество деревьев 6 Точность 0.7838258164852255
Количество деревьев 7 Точность 0.7807153965785381
Количество деревьев 8 Точность 0.7869362363919129
Количество деревьев 9 Точность 0.7947122861586314
Количество деревьев 10 Точность 0.7931570762052877
Количество деревьев 11 Точность 0.7869362363919129
Количество деревьев 12 Точность 0.7978227060653188
Количество деревьев 13 Точность 0.7900466562986003
Количество деревьев 14 Точность 0.7931570762052877
Количество деревьев 15 Точность 0.7916018662519441
Количество деревьев 16 Точность 0.7884914463452566
Количество деревьев 17 Точность 0.7916018662519441
Количество деревьев 18 Точность 0.7916018662519441
Количество деревьев 19 Точность 0.7947122861586314
Количество деревьев 20 Точность 0.799377


<div class="alert alert-info">
<b>🎯 Комментарий : </b> 
<br> В данном случае: Количество деревьев 21; Лучшая точность 0.80248833592535.
    
</div>

In [ ]:
    best_accuracy = 0



    for depth in range(1, 60):
        model = RandomForestClassifier(random_state=12345, n_estimators=21,max_depth = depth)
        model.fit(features_train,target_train)
        predictions_valid = model.predict(features_valid) 
        result = accuracy_score(target_valid,predictions_valid)
        print('Количество деревьев', 21,'Точность',result,'Высота',depth)
        if result > best_accuracy:
            best_accuracy =result
            best_depth_rf = depth
    print('Количество деревьев', 21,'Лучшая точность', best_accuracy,'Лучшая высота',best_depth_rf)

Количество деревьев 21 Точность 0.7340590979782271 Высота 1
Количество деревьев 21 Точность 0.7993779160186625 Высота 2
Количество деревьев 21 Точность 0.8055987558320373 Высота 3
Количество деревьев 21 Точность 0.807153965785381 Высота 4
Количество деревьев 21 Точность 0.8087091757387247 Высота 5
Количество деревьев 21 Точность 0.8211508553654744 Высота 6
Количество деревьев 21 Точность 0.8118195956454122 Высота 7
Количество деревьев 21 Точность 0.8055987558320373 Высота 8
Количество деревьев 21 Точность 0.8118195956454122 Высота 9
Количество деревьев 21 Точность 0.8149300155520995 Высота 10
Количество деревьев 21 Точность 0.8195956454121306 Высота 11
Количество деревьев 21 Точность 0.8164852255054432 Высота 12
Количество деревьев 21 Точность 0.80248833592535 Высота 13
Количество деревьев 21 Точность 0.8040435458786936 Высота 14
Количество деревьев 21 Точность 0.8009331259720062 Высота 15
Количество деревьев 21 Точность 0.80248833592535 Высота 16
Количество деревьев 21 Точность 0.7884

In [ ]:
best_model_ForestClass = RandomForestClassifier(random_state=12345, n_estimators=21,max_depth = 6)
best_model_ForestClass.fit(features_train,target_train)
predictions_valid = best_model_ForestClass.predict(features_valid) 
result = accuracy_score(target_valid,predictions_valid)
print('Количество деревьев',21,'Лучшая точность', result)

Количество деревьев 21 Лучшая точность 0.8211508553654744


In [ ]:
model = RandomForestClassifier(random_state=12345, n_estimators=21,max_depth = 6,criterion="entropy")
model.fit(features_train,target_train)
predictions_valid = model.predict(features_valid) 
result = accuracy_score(target_valid,predictions_valid)
print('Количество деревьев', 21,'Точность',result)

Количество деревьев 21 Точность 0.8118195956454122




<div class="alert alert-info">
<b>🎯 Комментарий : </b> 
<br> В данном случае точность снижена.Поменяем другой гиперпараметр.
    
</div> 

In [ ]:
model = RandomForestClassifier(random_state=12345, n_estimators=21,max_depth = 6,bootstrap=False)
model.fit(features_train,target_train)
predictions_valid = model.predict(features_valid) 
result = accuracy_score(target_valid,predictions_valid)
print('Количество деревьев', 21,'Точность',result)

Количество деревьев 21 Точность 0.8040435458786936


 

<div class="alert alert-info">
<b>🎯 Комментарий : </b> 
<br> Данный гиперпараметр не повлиял на результат.
    
</div> 

In [ ]:
model = RandomForestClassifier(random_state=12345, n_estimators=21,max_depth = 6,warm_start=True)
model.fit(features_train,target_train)
predictions_valid = model.predict(features_valid) 
result = accuracy_score(target_valid,predictions_valid)
print('Количество деревьев', 8,'Точность',result)

Количество деревьев 8 Точность 0.8211508553654744



    
<div class="alert alert-info">
<b>🎯 Комментарий : </b> 
<br> Точность снова стала лучше.
    
</div> 

In [ ]:
best_max_iter = 0
best_accuracy = 0

for max_iter in range(1,201):
    model = LogisticRegression(random_state=23042020, max_iter=max_iter)
    model.fit(features_train,target_train)
    predictions_valid = model.predict(features_valid) 
    result = accuracy_score(target_valid,predictions_valid)
    #print('Количество итерраций', max_iter,'Точность',result)
    if result > best_accuracy:
        best_max_iter = max_iter
        best_accuracy = result

print('Лучшая количество итерраций', best_max_iter,'Лучшая точность', best_accuracy)  

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Лучшая количество итерраций 77 Лучшая точность 0.7573872472783826


  
<div class="alert alert-info">
<b>🎯 Комментарий : </b> 
<br>Лучшее количество итерраций 77; Лучшая точность 0.7573872472783826.
    
</div> 

In [ ]:
best_model = LogisticRegression(random_state=12345,max_iter=77)
best_model.fit(features_train,target_train)
predictions_valid = best_model.predict(features_valid) 
result = accuracy_score(target_valid,predictions_valid)
print('Количество итерраций', 77,'Точность',result)

Количество итерраций 77 Точность 0.7573872472783826


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


In [ ]:
best_solver = ''
best_accuracy = 0

for solve in {'newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'}:
    model = LogisticRegression(random_state=12345)
    model.fit(features_train,target_train)
    predictions_valid = model.predict(features_valid) 
    result = accuracy_score(target_valid,predictions_valid)
    print('Solver', solve,'Точность',result)
    if result > best_accuracy:
        best_solver= solve
        best_accuracy = result

print('Лучший solver', best_solver,'Лучшая точность', best_accuracy)

Solver liblinear Точность 0.7465007776049767
Solver lbfgs Точность 0.7465007776049767
Solver newton-cg Точность 0.7465007776049767
Solver sag Точность 0.7465007776049767
Solver saga Точность 0.7465007776049767
Лучший solver liblinear Лучшая точность 0.7465007776049767


In [ ]:
model = LogisticRegression(random_state=12345,max_iter= 1000,tol=1e-6,solver = 'lbfgs')
model.fit(features_train,target_train)
predictions_valid = model.predict(features_valid) 
result = accuracy_score(target_valid,predictions_valid)
print('Точность',result)

Точность 0.7465007776049767


<div class="alert alert-info">
<b>🎯 Комментарий : </b> 
<br>Самая лучшая точность при  solver lbfgs. При изменение других гиперпараметров точность не меняется..
    
</div> 

## Проверьте модель на тестовой выборке

In [ ]:
test_predictions = best_model_TreeClass.predict(features_test)
accuracy = accuracy_score(target_test, test_predictions)
print('Точность модели решающего леса на тестовой выборке',accuracy)

Точность модели решающего леса на тестовой выборке 0.7853810264385692


In [ ]:
test_predictions = best_model_ForestClass.predict(features_test)
accuracy = accuracy_score(target_test, test_predictions)
print('Точность модели случайного леса на тестовой выборке',accuracy)

Точность модели случайного леса на тестовой выборке 0.8009331259720062


In [ ]:
test_predictions = best_model.predict(features_test)
accuracy = accuracy_score(target_test, test_predictions)
print('Точность модели логистической регресиии на тестовой выборке',accuracy)

Точность модели логистической регресиии на тестовой выборке 0.7402799377916018


<div class="alert alert-info">
<b>🎯 Комментарий : </b> 
<br>У всех моделей также высокая точность, но разница между тестовой и валидационной модели решающего леса и случайного леса где-то на 0.2. А модель логистической регресиии на тестовой выборке всего на 0.1. Т.е, переучение данной модели не грозит.
    
</div> 

## (бонус) Проверьте модели на адекватность

In [ ]:
test_predictions = best_model_ForestClass.predict(features_test)
accuracy = accuracy_score(target_test, test_predictions)
precision = precision_score(target_test, test_predictions)
recall = recall_score(target_test, test_predictions)
print('Accuracy =', accuracy, 'Precision =', precision, 'Recall =', recall)

Accuracy = 0.8009331259720062 Precision = 0.776 Recall = 0.49238578680203043


In [ ]:
f_score = f1_score(target_test, test_predictions)
print('F-мера =', f_score)

F-мера = 0.6024844720496894
